man1/uconv.1
.\" Copyright (C) 2013-2026 Kevin Boone and others
.\" Permission is granted to any individual or institution to use, copy, or
.\" redistribute this software so long as all of the original files are
.\" included, and that this copyright notice
.\" is retained.
.\"
.TH uconv 1 "June 2026"
.SH NAME
uconv \- a general-purpose unit converter 
.SH SYNOPSIS
.B uconv
.RB [options]\ {value}{from_units}\ {to_units}

.B uconv 
.RB [options]\ {value}\ {from_units}\ {to_units} 
.PP

.B uconv
.RB [options]\ -m\ {value}{from_units}...\ {to_units}
.PP

.SH DESCRIPTION
\fIuconv\fR is 
a general-purpose unit converter for use on the 
command line. There are many (perhaps too many) unit converter programs, but 
most have graphical interfaces. \fIuconv\fR is sufficiently
flexible in its unit specification that it should be faster and easier to
use on the command line than a graphical program. However, to use  \fIuconv\fR 
effectively, you do need to be familiar with the way it parses unit names.

\fIuconv\fR is primarily intended for domestic and general scientific use --
it doesn't include the large numbers of specialized units that 
particular disciplines employ.  

\fIuconv\fR recognizes a hundred or so basic units. These can be used
in combination to form compound units ('metres/second', 'newton.metres', etc). 
In fact, \fIuconv\fR will convert between any pair of units that are
dimensionally consistent. For example, you can convert 'joules per second per
kilogram' to 'calories per minute per pound', if you wish. The program won't
attempt to convert, for example, minutes to metres: these units are not
dimensionally consistent. 

Units that are reciprocal to one another are assumed to be dimensionally
consistent, and \fIuconv\fR will invert the result according. So, for example,
it can convert 'miles/gallon' to 'litres/km', even though these units are
reciprocal, rather than equivalent. 

Although the examples in this manual are mostly shown using 3 arguments -- one for the
source number, one for its units, and one for the target units -- the source units can also be
specified in the same argument as number.

.SH OVERVIEW OF OPERATION

The basic command is:

.nf
uconv {value} {from_units} {to_units}
.fi

Or, for the two argument form:

.nf
uconv {value}{from_units} {to_units}
.fi

For example, to convert 6 feet to centimetres:

.nf
uconv 6 ft cm
.fi

And the equivalent two-argument form:

.nf
uconv 6ft cm
.fi

\fIuconv\fR also accepts fractional values:

.nf
uconv 1/32in mm
uconv "7 1/2mi" km
uconv "-20 1/2" f c
.fi

When using mixed fractions in the form of 'a b/c' or '-a b/c', neither the
numerator nor the denominator can be negative, and the numerator must be less
than the denominator.

Because \fIuconv\fR recognizes a large number of synonyms for 
unit names, the following all produce the same result:

.nf
uconv 6 foot cm
uconv 6 feet centimetres
uconv 6 ft centimeter
.fi

Units can be compounded, provided they can be reduced to the same
base units. For example, to convert 60 miles per hour to metres per second:

.nf
uconv 60 miles/hour m/sec
.fi

The following are alternative ways to write the same conversion, and
all produce the same result:

.nf
uconv 60 mph metre/sec
uconv 60 mile.hour^-1 m/s 
uconv 60 miles.hr-1 m.second-1
.fi

Note that '^-1' is the conventional way of writing powers in programming
languages. This can be shortened just to '-1' in \fIuconv\fR for
convenience. Similarly, square meters can be written:

.nf
sqm
m^2
m2
"square meters"
.fi

The quotes around the last example are necessary to prevent the shell
recognizing the space as an argument separator.

There are two units representing angular degrees in uconv: 'degrees' and 'dms'. These
are internally equivalent but, when using 'dms' as the conversion target,
\fIuconv\fR displays the output with the fractional part
subdivided into arc minutes and arc seconds. 'degrees', however, always
produces a decimal result. 
Using '-d' with 'dms' is the same as using 'degrees':

.nf
$ uconv 1.7 radians degrees
1.7 radians = 97.4028 degrees
$ uconv 1.7 radians dms
1.7 radians = 97 degrees, 24 arc minutes, 10.1706 arc seconds
$ uconv -d 1.7 radians dms
1.7 radians = 97.4028 degrees
.fi

Multiple inputs can be specified with the '-m' option:

.nf
$ uconv -m 1ft 5yards m
1 foot = 0.3048 metres
5 yards = 4.572 metres
.fi

When specifying the units for the inputs, they must always be concatenated with
the value. When units are omitted, the last explicitly specified units are used
for the following values:

.nf
$ uconv -m 1ft 2 30cm 150 m
1 foot = 0.3048 metres
2 feet = 0.6096 metres
30 centimetres = 0.3 metres
150 centimetres = 1.5 metres
.fi

.SH UNIT FORMAT

A unit is made up of one or more unit elements separated by '.' or '/'. For
example:

.nf
N.m
J/sec
.fi

In books it's quite common to see newton-metres written 'nM'; \fIuconv\fR does
not recognize this notation, because it makes it too easy to confuse
single units with collections of units. For example, is 'msec' metre.seconds or
milliseconds? Because of this ambiguity, \fIuconv\fR insists on a separator
between unit elements: 'n.M' or 'newton.metres'.


The general format of a unit element is:

.nf
[dimension_name][prefix]{name}[[^]{[-]{dimension_number}]}
.fi

.SS Dimensions

The dimensions of the unit are its exponent, or a prefix indicating
the exponent, but not both. For example, cubic centimetres can be written

.nf
cucm
"cu cm"
"cubic cm"
cm3
cm^-3
.fi

Similarly, the second power can be represented as '^2' or '2' after the unit name,
or 'square' or 'sq' before it. Higher powers can be entered using numbers, but
there are no names for these.

.SS Prefixes

Unit names can be preceded by SI prefixes: tera(T) giga(G), mega(M),
kilo(k), deci(d), centi(c), milli(m), micro(u), nano(n), pico(p).

\fIuconv\fR also recognizes IEC prefixes for data capacity units
(gibibytes, mebibytes). These prefixes represent power-of-two
multipliers, rather than the power-of-ten multipliers used for all
other units. IEC prefixes are only allowed for units involving data
capacity: there is no 'gibigram', for example. By default,
abbreviations are treated as IEC units unless the \fI-s\fR switch is
given or abbreviations are mixed e.g. "20 mb mib".

Prefixes are case-sensitive, although units generally are not.
An exception is 'mb' for 'megabyte' -- see below for more on this point.

.SS Names

\fIuconv\fR accepts units with a variety of names and abbreviations. In general,  
\fIuconv\fR's name match is case insensitive even where the conventional name 
is upper-case or mixed case (e.g., mmHg for millimetres of mercury). 

Plural forms (feet, rather than feet) are always recognized even when
they are grammatically questionable ('1 feet', '2 foot').

Some unit names have spaces (e.g., nautical miles, troy ounces). \fIuconv\fR 
provides
versions of these names with hyphens (nautical-miles, troy-ounces), because the 
spaces
tend to confuse the shell. Of course, you can still use the correct names 
with spaces, but you'll probably have put them in quotation marks.

To see a list of all unit names and synonyms, use \fIuconv -l\fR.


.SH OUTPUT FORMAT

\fIuconv\fR displays results to five significant figures, using scientific
notation where necessary. Where groups of units form sets with a non-decimal
relationship, the default behaviour is to split the result into
elements of the set. For example:

.nf
$ uconv 1000 l gal
1000 litres = 219 gallons, 7 pints, 12.0638 fluid-ounces
.fi

This behaviour is used for tons/hundredweight/stones/pounds/ounces, 
hours/minutes/seconds, and gallons/pints/fluid-ounces. To override
this formatting style and use only decimal formatting, use the \fI-d\fR
switch:

.nf
$ uconv -d 1000 l gal
1000 litres = 219.969 gallons
.fi

The output includes the input units, but with full names rather than any 
abbreviations you might have used. This is necessary because, with
such a large number of units available, it's very easy to use the wrong
abbreviation (entering 'C' for coulomb, when \fIuconv\fR interprets it 
as Celsius, for example). You should always check that the units in the output are what
you expect, and refer to the list of units if necessary.


.SH THINGS TO WATCH OUT FOR

In its output, \fIuconv\fR favours traditional spellings of 
unit names: 'gramme' 
over 'gram', 'metre' over 'meter'. Of course, both forms are accepted
as input.

Kilogrammes, pounds, etc., are units of mass, not weight. \fIuconv\fR has
to make this distinction, because otherwise it's difficult to ensure
that consistent units are being converted. The distinction is not
always noticeable: plain 'weights' in 
kilograms and stones can be freely converted -- you're actually converting
masses, but the conversion factors are the same, so you get the correct
result. What you can't do is
convert, for example, torque in newton-metres to pounds-feet,
because a pound is a mass, not a force. 
The associated unit of torque is pound-force-feet.
Consequently, we can't do this:

.nf
uconv 1 lb.foot N.m
.fi

But we can do this:

.nf
uconv 1 lbf.foot N.m
.fi

Because 'lbf' is the abbreviation for pounds-force. This is an annoyance, because 
we rarely use the suffix '-force' outside of physics textbooks. But 
allowing \fIuconv\fR to be flexible here would potentially cause other problems
in parsing.

If an argument to \fIuconv\fR will match a unit name alone, or a unit with a prefix,
then it prefers the unit name alone. So, for example, 'min' is 
interpreted as 'minute', and not 'milli-inch'. This is usually the right
thing to do. If \fIuconv\fR matches wrongly in this respect, give the
prefix in full ('milliinch').

Sometimes the same name is used for different units. There isn't much that
\fIuconv\fR can do about this. For example, the unit 'cord' is used as a unit
of area (mostly in the UK), and of volume (mostly in North America). The
same is true for abbreviations -- 'C' is used for both coulombs and Celsius. 
There's no
way to get around this problem other than providing different name or abbreviation
variants in
the program. Refer to the list of unit names to see which is appropriate.

Where there is a regional variation, \fIuconv\fR favours English units over
US. For example, a US pint is slightly smaller than a UK pint
(but there are still eight pints to the gallon in both regions). \fIuconv\fR
uses the prefix 'us' to indicate US variants, e.g., 'uspint', 'usgallon', 'uston'. 

Although it's not very mathematically sound, \fIuconv\fR recognizes 'kilo'
as an abbreviation for 'kilogramme'. Of course, 'kilo' is really a prefix;
none of the other prefixes (mega, centi, etc) are interpreted in this
loose way. 'kilo' is handled differently because its use as a measure of
mass is so common.

The output of \fIuconv -l\fR doesn't show prefixed units, even when they are
very common. So, while you can use 'km' as a unit of length, it doesn't show
up in the unit list. The reason for this is that there are so many units that
including each one with all possible prefixes would make the unit list unreadably.
long.

.SH NOTES ON SPECIFIC UNITS

\fIuconv\fR uses the name 'calorie' (or 'cal') in the scientific sense:
the amount of energy that will raise the temperature of
one gramme of water by one degree Celsius. The 'calorie' used
in food labels and diet books is typically a thousand times 
bigger than this -- strictly
it's a kilocalorie. Food labelling is getting better in the UK, but vendors still
often use the term 'calorie' to mean a kilocalorie. You can 
use 'kcal' to denote the food-related use of the calorie. 

The units of radiation dose -- gray, rad, sievert, and REM -- can be
freely -- and possibly wrongly -- converted using 
\fIuconv\fR because they have the same dimensions.
However, in practice they are not convertible, except where the radiation
is photonic. This is an issue of physics, not mathematics.

Radians cannot be abbreviated to 'rad' or 'rads', as these names are
used by the unit of radiation dose, the rad. You need to write 'radians'.

The steradian is a dimensionless unit of solid angle, which cannot be 
converted to anything else. It exists in \fIuconv\fR solely to prevent units
of illuminance being converted to units of luminance. Luminance is defined
in terms of luminous intensity, while illuminance is defined in terms of
luminious flux. These units are generally not convertible.

The units 'footcandle' and 'footlambert' are common, but misleading, names
for 'lumens/sq foot  and  candela/sq foot  respectively. These units cannot
be entered as 'foot.candela' or 'foot.lambert' -- these are valid, but completely
different -- units.

\fIuconv\fR takes the units 'candela' and 'candlepower' to be equivalent.
Candlepower has been expressed in different ways in the past, but for the
last fifty years or so it's been taken to be equivalent to the candela
in most disciplines. 

The unit 'mpg', which is a shortcut for 'miles/gallon' refers to the UK
(imperial) gallon. For US gallons, use 'miles/usgallon'.
 
The units for digital storage and transmission have explicit definitions for
the various prefixes so they can be entered in lower-case even though this
usage violates SI
conventions. For example, "1 mbyte" will be interpreted as "1 megabyte" rather
than "1 millibite". This usage is allowed because it's so common, and
shouldn't cause problems in practice.

.SH "OPTIONS"
.TP
.BI -h
Show brief usage information 
.LP
.TP
.BI -l
Print a list of unit names and synonyms
.LP
.TP
.BI -s
Use powers of 10 (SI) instead of 2 (IEC) for bytes and bits. Normally something
like "1 kb" would be interpreted as 1024 bytes instead of 1000 bytes. If the input
contains a mixture of SI prefixes and IEC prefixes (e.g. "20 mb mib"), \fIuconv\fR presumes
that you 
want to convert between SI and IEC units.
.LP
.TP
.BI -v
Show version number and exit
.LP

.SH EXAMPLES

.nf
$ uconv 100 mph kmh
100 mph = 160.934 kmh

$ uconv 10 l/s gal/h
10 litre/second = 7918.89 gallon/hour

$ uconv 8 cuft l
8 foot^3 = 226.535 litre

$ uconv -270 c k
-270 celsius = 3.15 kelvin

$ uconv 5000 sqm acre
5000 metre^2 = 1.23553 acre

$ uconv 1 btu W.s
1 BTU = 1054.5 watt.second

$ uconv 3705 sec hr
3705 seconds = 1 hour, 1 minute, 9 seconds

$ ./uconv 104 kilos st
104 kilos = 16 stones, 5 pounds, 4.49204 ounces

$ uconv -m 5lbs 2 16oz kg
5 pounds = 2.26796 kilogrammes
2 pounds = 0.907185 kilogrammes
16 ounces = 0.453592 kilogrammes
.fi 


.SH BUGS AND LIMITATIONS

\fIuconv\fR is strongly biased to US and UK unit names, and particularly to the
UK.  It recognizes and displays only English, and would not be easy to extend
to other languages.

Although \fIuconv\fR can output imperial measures in non-decimal form 
(pounds and ounces, etc), it cannot read data in this format. Input
values have to be decimal.

\fIuconv\fR tries to be correct in its grammar, that is, to output
values like "1 foot, 2 inches" rather than "1 feet, 2 inches". However,
when a value comes from a computation, it may be inexact. So, for example,
if the result is 1.0000001 feet, it will likely be displayed as
"1 feet", because 1.0000001 is not exactly 1. Problems like this are
tricky to fix without re-implementing the entire math library to work
in decimal.

Some relatively common units (e.g., 'barrels') are not included because 
there are too many
variants to be sure of getting the right value.

On the whole, \fIuconv\fR includes units in its database only if they are in
current use, and avoids those of only historical significance ('chain', 'ell',
 'league'). These units could easily be included, but doing so increases the risk
of the user entering a partial name or abbreviation that matches something
completely unexpected. In general, the more units in the database, the more
likely such inadvertent misreadings becomes.

\fIuconv\fR does not recognize 'per' as a compound unit separator. That is,
you can say 'km/hr', but not 'km per hour'.  
 


.SH AUTHOR AND LEGAL
\fIuconv\fR
is maintained by Kevin Boone, with significant contributions from
other authors, and is open source under the
terms of the GNU Public Licence, version 3.0. There is no warranty
of any kind.




.\" end of file


uconv.c
/*============================================================================
  uconv.c

  (c)2013-2021 Kevin Boone
  Distributed under the terms of the GNU Public Licence, version 2
============================================================================*/

// Must define _GNU_SOURCE for strcasestr() to work properly
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include "units.h" 

static BOOL default_to_iec = TRUE;
static BOOL force_decimal = FALSE;

typedef enum {
  no_prefix,
  iec_prefix,
  si_prefix,
  digital_storage_prefix_enum_count,
} DigitalStoragePrefixType;


/*============================================================================
  show_version 
============================================================================*/
void show_version (void)
  {
  printf ("%s version %s\n", NAME, VERSION);
  printf ("Copyright (c)2013-2026 Kevin Boone and others\n");
  printf ("Freely distrbutable under the terms of the GNU Public Licence, v3.0");
  printf ("\n");
  }


/*============================================================================
  show_usage
============================================================================*/
void show_usage (const char *argv0, FILE *out)
  {
  fprintf (out, "Usage: %s [options] {number} {from_units} {to_units}\n", argv0);
  fprintf (out, "Options:\n");
  fprintf (out, "  -d                Force decimal output\n");
  fprintf (out, "  -h                Show this message\n");
  fprintf (out, "  -l                List available units\n");
  fprintf (out, "  -m                Accept multiple input values\n");
  fprintf (out, "  -s                Use powers of 10 instead of 2 for bytes and bits\n");
  fprintf (out, "  -v                Show version\n");
  }


/*============================================================================
  data_unit_type
============================================================================*/
int data_unit_type (Unit u)
  {
  switch (u)
    {
    case kibibyte:
    case mebibyte:
    case gibibyte:
    case tebibyte:
    case pebibyte:
    case exbibyte:
    case kibibit:
    case mebibit:
    case gibibit:
    case tebibit:
    case pebibit:
    case exbibit:
      return iec_prefix;

    case kilobyte:
    case megabyte:
    case gigabyte:
    case terabyte:
    case petabyte:
    case exabyte:
    case kilobit:
    case megabit:
    case gigabit:
    case terabit:
    case petabit:
    case exabit:
      return si_prefix;

    default:
      return no_prefix;
    }
  }


/*============================================================================
  si_to_iec
============================================================================*/
int si_to_iec (Unit u)
  {
  switch (u)
    {
    case kilobyte: return kibibyte;
    case megabyte: return mebibyte;
    case gigabyte: return gibibyte;
    case terabyte: return tebibyte;
    case petabyte: return pebibyte;
    case exabyte: return exbibyte;
    case kilobit: return kibibit;
    case megabit: return mebibit;
    case gigabit: return gibibit;
    case terabit: return tebibit;
    case petabit: return pebibit;
    case exabit: return exbibit;
    default: return u;
    }
  }

/*============================================================================
  has_explicit_si_data_unit
  Detect whether an SI data unit is spelled out in full.
============================================================================*/
BOOL has_explicit_si_data_unit (char *units)
  {
  char *prefixes[] = {
    "kilob", "megab", "gigab", "terab", "petab", "exab", NULL,
  };

  for (char **cursor = prefixes; *cursor; cursor++)
    {
    size_t prefix_length = strlen(*cursor);
    char *match = strcasestr(units, *cursor);


    if (match)
      {
      char next_char = tolower((int) *(match + prefix_length));

      // We're abusing the fact that data units are the only ones that start
      // with a "b" and are followed by a "y" or "i".
      if (next_char == '\0' || strchr("yi./", next_char))
        return TRUE;
      }
    }

  return FALSE;
  }

/*============================================================================
  fractod
  Like strtod(3) but also handles fractions in the form of "a/b" and "a b/c".
============================================================================*/
double fractod (char *text, char **endptr)
  {
  double a, b, result = 0;
  int length = 0, x;

  if (sscanf(text, "%d %lf / %lf %n", &x, &a, &b, &length) != EOF && length)
    {
    if (a < 0 || b <= 0 || a > b)
      {
      length = 0;
      errno = EINVAL;
      }
    else
      {
      result = (double) x + (x < 0 ? -a / b : a / b);
      errno = 0;
      }
    }
  else if (sscanf(text, "%lf / %lf %n", &a, &b, &length) != EOF && length)
    {
    if (b <= 0)
      {
      length = 0;
      errno = EINVAL;
      }
    else
      {
      result = a / b;
      errno = 0;
      }
    }
  else if (sscanf(text, "%lf %n", &a, &length) != EOF && length)
    {
    result = a;
    errno = 0;
    }
  else if (*text == '\0')
    errno = EINVAL;

  if (endptr)
    *endptr = text + length;

  return result;
  }


/*============================================================================
  convert
  Perform a conversion of one unit to another. If the value and units are
  separate strings, they are passed in as "from" and "from_units_suffix",
  respectively. If they are concatenated, "from" should point to the whole
  string while "from_units_suffix" should be set to NULL.
============================================================================*/
int convert (char *from, char *from_units_suffix, char *to)
  {
  static char* previous_from_units_suffix = NULL;

  double value;
  char *error = NULL, *invalid = NULL;
  errno = 0;
  Units *fu = NULL, *tu = NULL;

  if (from_units_suffix)
    {
    value = fractod (from, &invalid);
    // If fractod parsed the entire string, ensure "invalid" is set to NULL.
    if (errno == 0 && invalid && *invalid == '\0') invalid = NULL;
    }
  else
    {
    value = fractod (from, &from_units_suffix);
    if (errno != 0 || from == from_units_suffix)
      {
      invalid = from;
      if (from != from_units_suffix) *from = '\0'; // Don't include units in error.
      }
    else if (*from_units_suffix == '\0')
      {
      if (!previous_from_units_suffix)
        {
        fprintf (stderr, "No units specified for input value '%s'\n", from);
        return 1;
        }

      // If the "from" value does not include units but a previous call did, we
      // reuse the units from the previous call.
      from_units_suffix = previous_from_units_suffix;
      }
    }

  if (invalid)
    {
    fprintf (stderr, "%s: %s\n", invalid, errno == 0 ? "Not a valid number" : strerror(errno));
    return 1;
    }

  fu = units_parse (from_units_suffix, &error);
  if (!fu) goto done;

  tu = units_parse (to, &error);
  if (!tu) goto done;

  // When defaulting to IEC units, only convert to IEC units if all inputs are
  // abbreviated SI units. This allows conversion of SI to IEC by mixing unit
  // types e.g. "10 gb gib". The logic for determining if a prefix is spelled
  // out is triggered by any prefix in the string rather than associating it
  // with the right unit. Specify multiple data units is an unusual case, so
  // this shouldn't be a problem in practice.
  if (default_to_iec)
    {
    int i, counts[digital_storage_prefix_enum_count] = {0};

    for (i = 0; i < fu->n_elements; i++)
      {
      int type = data_unit_type (fu->units[i].unit);
      if (type != si_prefix || !has_explicit_si_data_unit(from_units_suffix)) counts[type]++;
      }

    for (i = 0; i < tu->n_elements; i++)
      {
      int type = data_unit_type (tu->units[i].unit);
      if (type != si_prefix || !has_explicit_si_data_unit(to)) counts[type]++;
      }

    if (counts[si_prefix] && !counts[iec_prefix])
      {
      for (i = 0; i < fu->n_elements; i++)
        fu->units[i].unit = si_to_iec (fu->units[i].unit);

      for (i = 0; i < tu->n_elements; i++)
        tu->units[i].unit = si_to_iec (tu->units[i].unit);
      }
    }

  double res = units_convert (value, fu, tu, &error);
  if (!error)
    {
    previous_from_units_suffix = from_units_suffix;
    char *fs = units_format_string_and_value (fu, value, force_decimal);
    char *ts = units_format_string_and_value (tu, res, force_decimal);
    printf ("%s = %s\n", fs, ts);
    // Oops -- not freeing these leads to a memory leak. How long
    //   has this bug been here? :/
    free (fs);
    free (ts);
    }

done:
  if (error) fprintf (stderr, "Error: %s\n", error);

  free(fu);
  free(tu);
  free(error);
  return error ? 1 : 0;
  }

/*============================================================================
  main
============================================================================*/
int main (int argc, char **argv)
  {
  int i, optind = 1;
  BOOL usage = FALSE;
  BOOL list = FALSE;
  BOOL version = FALSE;
  BOOL multiple_inputs = FALSE;

  // We have to parse the arguments manually, because the first argument
  //  might be a negative number
  for (i = 1; i < argc; i++)
    {
    if (argv[i][0] == 0) continue;
    if (argv[i][0] == '-')
      {
      if ((int)strlen (argv[i]) > 1)
        {
        if (!isdigit ((int)argv[i][1]))
          {
          int j, l = strlen (argv[i]);
          for (j = 1; j < l; j++)
            {
            switch (argv[i][j])
              {
              case 'd':
                force_decimal =TRUE;
                break;
              case 's':
                default_to_iec =FALSE;
                break;
              case 'v':
                version =TRUE;
                break;
              case 'l':
                list =TRUE;
                break;
              case 'm':
                multiple_inputs =TRUE;
                break;
              case 'h':
                usage =TRUE;
                break;
              }
            }
          optind++;
          }
        }
      }
    else
      break;
    }

  if (usage)
    {
    show_usage (argv[0], stdout); 
    exit(0);
    }

  if (version)
    {
    show_version (); 
    exit(0);
    }
  
  if (list)
    {
    units_dump_tables (stdout); 
    exit(0);
    }

  if (!multiple_inputs)
    {
    switch (argc - optind)
      {
      case 2:
        return convert(argv[optind], NULL, argv[optind + 1]);
      case 3:
        return convert(argv[optind], argv[optind + 1], argv[optind + 2]);
      default:
        fprintf (stderr, "%s: Wrong number of arguments; expected 2 or 3\n",
          argv[0]);
        show_usage (argv[0], stderr);
        return 1;
      }
    }
  else if ((argc - optind) < 2)
    {
    fprintf (stderr, "%s: Wrong number of arguments for use with -m; expected at least 2\n", argv[0]);
    return 1;
    }
  else
    {
    int status = 0;

    for (int i = 0; i < argc - optind - 1; i++)
      status |= convert(argv[optind + i], NULL, argv[argc - 1]);

    return status;
    }
  }
